src/serializer/utils.js

```diff
@@ -249,3 +253,101 @@ export function handleReportedSideEffect(
     exceptionHandler(`side-effects from throwing exception${location}`, expressionLocation);
   }
 }
+
+function isDefinedInsideFunction(
+  childFunction: FunctionValue,
+  maybeParentFunctions: Set<FunctionValue>,
+  optimizedFunctionsAndEffects: Map<FunctionValue, AdditionalFunctionEffects>
+): boolean {
+  for (let maybeParentFunction of maybeParentFunctions) {
+    if (childFunction === maybeParentFunction) {
+      continue;
+    }
+    // for optimized functions, we should use created objects
+    let maybeParentFunctionInfo = optimizedFunctionsAndEffects.get(maybeParentFunction);
+    if (maybeParentFunctionInfo && maybeParentFunctionInfo.effects.createdObjects.has(childFunction)) return true;
+    else {
+      // for other functions, check environment records
+      let env = childFunction.$Environment;
+      while (env.parent !== null) {
+        let envRecord = env.environmentRecord;
+        if (envRecord instanceof FunctionEnvironmentRecord && envRecord.$FunctionObject === maybeParentFunction)
+          return true;
+        env = env.parent;
+      }
+    }
+  }
+  return false;
+}
+
+// Check if an optimized function defines the given set of functions.
+function definesFunctions(
+  possibleParentFunction: FunctionValue,
+  functions: Set<FunctionValue>,
+  optimizedFunctionsAndEffects: Map<FunctionValue, AdditionalFunctionEffects>
+): boolean {
+  let maybeParentFunctionInfo = optimizedFunctionsAndEffects.get(possibleParentFunction);
+  invariant(maybeParentFunctionInfo);
+  let createdObjects = maybeParentFunctionInfo.effects.createdObjects;
+  for (let func of functions) if (func !== possibleParentFunction && !createdObjects.has(func)) return false;
+  return true;
+}
+
+export type HelperState = {|
```

Comment:
This is not a good type name:
1) "Helper"?
2) "State": As far as I can tell, it just aggregates three _values_ that are not supposed to be mutated by anyone at this point.